{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import unicode_literals, print_function\n",
    "\n",
    "import plac\n",
    "import random\n",
    "from pathlib import Path\n",
    "import spacy\n",
    "from spacy.util import minibatch, compounding\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load the prepared data\n",
    "import pickle\n",
    "\n",
    "with open(\"D:/mlds_data/text_mining/projet/prepared_data/cleaned_data_V1.txt\", \"rb\") as fp:   # Unpickling\n",
    "\n",
    "    DATA = pickle.load(fp)\n",
    "    \n",
    "    #take the 100 last examples for test\n",
    "    TEST_DATA=DATA[len(DATA)-100:len(DATA)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading from D:\\mlds_data\\text_mining\\projet\\model\n"
     ]
    }
   ],
   "source": [
    "#load the saved model\n",
    "output_dir=\"D:/mlds_data/text_mining/projet/model\"\n",
    "output_dir = Path(output_dir)\n",
    "\n",
    "print(\"Loading from\", output_dir)\n",
    "nlp_test = spacy.load(output_dir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8328 -> 8333 GENE : BRCA1\n",
      "13097 -> 13102 GENE : BRCA1\n",
      "2044 -> 2051 GENE : CACNA1C\n",
      "2146 -> 2153 GENE : CACNA1C\n",
      "2306 -> 2313 GENE : CACNA1C\n",
      "2384 -> 2391 GENE : CACNA1C\n",
      "2457 -> 2464 GENE : CACNA1C\n",
      "2766 -> 2773 GENE : CACNA1C\n",
      "4813 -> 4815 GENE : GC\n",
      "1256 -> 1263 GENE : CYP2C19\n",
      "8371 -> 8377 DISEASE : Cancer\n",
      "1668 -> 1681 DISEASE : Schizophrenia\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div class=\"entities\" style=\"line-height: 2.5\">Untoward reactions to medications are a major cause of medical morbidity and mortality, and genetic tests for susceptibility to these reactions have entered into pharmaceutical regulations in many countries. The potential for other applications of pharmacogenomics, such as development of personalized medicine based on the patient's genetic susceptibility to disease, has so far been realized only for a few drugs, but a large number of such applications may be expected in the future. In this paper we endeavor to anticipate the research that will lead to these pharmacogenomics innovations, and the ethical and public policy challenges that may emerge in the development of this knowledge and in its application to public health. Decades ago, it was discovered that normal variants in metabolizing enzymes can lead to serious reactions to medications (drug-gene interactions; DGIs) in patients who metabolize medications much more slowly than average. DGIs can also cause lack of efficacy in patients who are rapid metabolizers. At this time, the most efficient way to determine susceptibility to these DGIs is by measuring the single-nucleotide polymorphisms (SNPs) that are associated with specific variants of the drug-metabolizing enzyme. CYP2D and \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    CYP2C19\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">GENE</span>\n",
       "</mark>\n",
       " genes. For most common diseases, pharmacodynamics, pharmacokinetics, and disease-related genetic association indications for clinically meaningful medication efficacy or for risk of adverse effects have not yet been found. The reason may lie in the current state of the genetics of common diseases. For most common diseases, most patients do not have any genetic association with a large effect on risk. \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    Schizophrenia\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">DISEASE</span>\n",
       "</mark>\n",
       " is typical. As of their July 2014 paper, The current state of SNP association with psychiatric disorders is that our knowledge of statistical genetic association with disease has advanced, but the development of pharmacogenomic molecular targets from this knowledge has been elusive. Common SNPs associated with disease frequently have known functional effects. \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    CACNA1C\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">GENE</span>\n",
       "</mark>\n",
       " and with which only a handful of patients survive past childhood, developed bipolar disorder. \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    CACNA1C\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">GENE</span>\n",
       "</mark>\n",
       " is the gene for Ca(v)1.2, a subunit protein of L-type calcium-channels expressed in brain, most strongly so in the cerebellum. A common polymorphism of \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    CACNA1C\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">GENE</span>\n",
       "</mark>\n",
       " has long been the most strongly associated SNP with pipolar disorder. \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    CACNA1C\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">GENE</span>\n",
       "</mark>\n",
       " in brain: patients with the risk allele have lower expression of CACNA1C. Currently, calcium channel blocking drugs (CCBs) are in wide use for cardiac arrhythmias and hypertension, and have previously been reported to have positive therapeutic effects in a proportion of patients with mania and depression, although the overall evidence for therapeutic effect is ambiguous. \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    CACNA1C\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">GENE</span>\n",
       "</mark>\n",
       " , that is, with the non-risk allele. This prediction was based on functional genomics of a common SNP, as well as the effect of a rare mutation that profoundly altered the gene in which that SNP is present; it is not based on the strength of a statistical association of the SNP on risk of disease. Similar reasoning was offered at the XXIth World Congress of Psychiatric Genetics, The multigenic nature of the major psychiatric disorders has been reinforced by findings of polygenic inheritance (ie, non-Mendelian inheritance) in schizophrenia, and by the finding that a given configuration of genetic markers cross multiple diagnostic boundaries. any de novo CNV mutation in the genome is associated with greatly increased risk for schizophrenia, bipolar disorder, and autism. The above observations and implications will need pathway and network approaches that parse the genetic associations with illness into groups of gene functions that are similar or related (pathways), In a molecular network, nodes represent molecules and edges represent relationships between the nodes; for example, the nodes may represent proteins and the edges protein-protein interactions. Network -based approaches stem from systems biology, a holistic approach to studying biological systems, in which disease-associated molecular changes are studied in their larger biological context, rather than in isolation. There is accumulating evidence that alterations in network structure are associated with neuropsychiatric disorders (for review, see ref 18) and psychotropic drug response. These molecular network approaches are still being developed; they have produced intriguing results, but have not yet led to any major breakthrough in neuropsychiatric disease studies. Network structure is highly context-dependent, meaning that networks vary through time and across tissues, cell types, brain layers, etc. Horvath and colleagues developed a mathematical algorithm for creating networks, which they called Weighted Gene Co-expression Network Analysis (WGCNA). NOTCH2 , a member of the Notch gene family, with the highest intramodular connectivity. This module is enriched with neuron development and neuron differentiation genes. The second module had metallothienin (MT) genes most prominently, with MTIX as the hub gene, Expression changes in these MT genes in schizophrenia and bipolar had been previously reported, In summary, many genes or gene products may be suitable targets for the pharmacological treatment of individuals with particular genetic dispositions. The ethical and public policy challenges these findings may present for research and public health progress through pharmacogenomics are worth discussing dispassionately now, when no practical application yet exists. Very-large-scale patient studies will be needed to detect and replicate drug responses and side effects beyond associations with drug-metabolizing enzymes. Participants in drug trials that collect blood but did not specify genetic tests, and patients who provide blood specimens to a hospital laboratory often have not consented to further research and testing on the specimens they provided (see review in ref 34). In the case of laboratory specimens (and autopsy specimens) consent is not considered required by current ethical standards in the United States, although serious issues would arise if an individual's identity were to be disclosed from the data. For prediction of therapeutic response and adverse events, it would obviously be very valuable to mine medical records and do genetic testing on leftover laboratory specimens on a very large scale. At present, data mining of hospital diagnostic and demographic records, where the individuals are made anonymous (ie, deidentified) is commonly done in research. Accessing medical test results is an informatics problem but not an ethical one. When consent is required, under current standards it cannot be a broad-based consent to any future research, and this is another barrier. It is impractical to recontact individuals for every additional study. For an individual for whom a significant number of genetic polymorphisms were assayed, the data itself could lead to unwanted identification, such as the well-known instances where it is possible to determine if an individual's genetic data was included in a genetic database on a particular disease, even if there were no personal identifiers in the disease database. Conflicting demands result, then, between medical research advances and the rights of an individual to privacy. The solution should lie in some combination of modifying the rules on consent to research to allow open-ended consent, continuing the policy of allowing research on existing laboratory and pathological specimens collected for medical testing, and establishing more vigorous deterrent penalties for violations of privacy. The United States Genetic Information Nondiscrimination Act of 2008 (Pub.L. 110-233, 122 Stat. 881), which prohibits the discrimination against individuals based on genetic information in matters involving health insurance and employment, might serve as a legal precedent and template for penalties on illegitimate use of an individual's genetic information. Biological patents that concern artificially modified genes are accepted around the world, but there are differences between jurisdictions on patenting a discovery of the function of a naturally occurring genetic variant. In Australia, the Federal Court in 2013 affirmed the Myriad corporation patent on the \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    BRCA1\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">GENE</span>\n",
       "</mark>\n",
       " variant as a test for breast cancer (\n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    Cancer\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">DISEASE</span>\n",
       "</mark>\n",
       " Voices Australia v. Myriad Genetics Inc. [2013] FCA 65), but the United States Supreme Court, ruling on the same test in 2013, found that naturally occurring DNA sequences are ineligible for patent protection (Association for Molecular Pathology v Myriad Genetics, U.S. Supreme Court, No. 12398). A utilitarian approach to ethics would weigh the benefits and costs on each side of such a decision. The benefit of no eligibility for a patent is that testing based on existing knowledge becomes cheaper, and the cost is that research into new information of this type becomes less economically worthwhile. The ultimate outcome of these issues is not predictable, but the unintended cost of inhibiting development of personalized medicine, under the US ruling, may be considerable. As discussed by Wertz, All of the approaches to treatment development discussed so far are built on quantitative approaches to genetics and genomics, usually quantitative associations based on large samples and thus applicable to large numbers of patients. However, a considerable proportion of individuals with psychiatric disorders, schizophrenia, bipolar, and autism spectrum disorder, have rare variants as their major risk component. The ethical challenges of these treatments include their costs, as well as the inadequacy of current procedures of pharmaceutical regulation to deal with questions for which there are no statistical samples of individuals to guide the regulators. For new drugs, the standard sequence of testing in humans is (in the United States): Phase 0: Pharmacodynamics and pharmacokinetics; Phase 1: Screening for safety in a small sample of human volunteers; Phase 2: Establishing the efficacy of the drug, usually against a placebo, in a limited sample of patients in a research setting; Phase 3: Final confirmation of safety and efficacy in a large sample; and Phase 4: Sentry studies after approval for medical use during sales to large numbers of patients. This sequence is appropriate for common and even some fairly uncommon diseases, but is obviously not appropriate for diseases associated with rare genetic variants. There may not be enough patients in the world to establish efficacy with statistical methods. Safety studies in normal volunteers or other patients with similar diseases may not provide information on risks directly related to a particular patient's genetic makeup. Yet, in the aggregate, rare and very rare diseases are common. Therefore, new general methods for testing personalized drugs are needed for this type of medicine to advance. One approach would be studies of the patient's rare variant, or of his or her entire genome, in suitable models. Neuron-like pluripotent stem cells can be developed from an adult patient's skin fibroblasts, The benefits of assaying individuals' drug metabolizing enzyme phenotypes to predict adverse drug reactions makes this area of pharmacogenomics uncontroversial. First, specific side effects related to genotype can be identified, and thus risk is reduced. Second, specific indications for use of a treatment can be identified from genetic testing. As a result, more drugs could theoretically be approved by regulatory agencies, with requirements or recommendations for genetic testing of individuals receiving the drug. Wertz has proposed routine testing for drug metabolizing enzymes as a public health preventive measure. The public enthusiasm for personalized medicine began in 1998, when trastuzumab was approved for treatment of breast cancer in cases where the tumor is overexpressing the HER2 protein. Currently, the FDA lists a considerable number of treatments for which pharmacogenomic biomarkers are recommended. Approximately 12% of drugs licensed in the period 1998-2012 had PGx (Pharmacogenomic) biomarker information included in their labels at the time of their approval. Of that number, labels direct clinicians to utilize PGx testing prior to prescribing treatments in only 14 cases. This clearly falls short of expectations many had in the 1990s about the transformative impact, of PGx. In most cases, the inclusion of this information currently has limited or no direct clinical utility. FDA approval of genetic test indications on the label has been considered a frustrating process by many investigators, and has occurred in many cases after the period of patent protection had expired. The cost of custom drug development for rare variants is another barrier to accessibility of genomic technologies in pharmaceuticals development. Wertz There are multiple ethical issues in genetic testing for disease markers that may well carry over into pharmacogenomics-based treatment. We have reviewed these issues elsewhere, \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em; box-decoration-break: clone; -webkit-box-decoration-break: clone\">\n",
       "    BRCA1\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; text-transform: uppercase; vertical-align: middle; margin-left: 0.5rem\">GENE</span>\n",
       "</mark>\n",
       " variant testing of a child for breast cancer susceptibility. When choice of treatment would be guided by testing, this is no longer an issue; the parent or guardian can reasonably give consent for testing a child who is under age, because of the potential benefit to the child who is tested. Within families, conflicts over testing for high-risk genomic markers can arise, because of concern over implications of test results for persons who have not been tested, and because of stigma within families and within society. In the discussion above, we anticipated that the use of stem cells and transgenic animal models may become more extensive for efficacy and toxicity analyses when a disease is rare and there is no statistical basis for conducting large-scale human studies. Apart from ethical controversies over experimental methods, we would also anticipate ethical and social policy controversy over the amount of risk a patient might reasonably undertake, or be advised to undertake, for a drug that cannot be tested for toxicity as extensively as drugs for common diseases. Lastly, there have been issues raised over protecting genetic information in a medical record, or from disclosure to insurance companies or employers. These are not issues specific for clinical pharmacogenomics, and we do not discuss them here, but they will become pertinent as this type of diagnosis and treatment matures. To summarize, there is a vast potential for pharmacogenomics applications to treatment of psychiatric disorders, but it has only recently become clear that this can be realized. The ethical and public policy issues do not appear to be insurmountable, or uniquely different from other issues of medical testing, such as infectious diseases where other family members or associates may be at risk. The technological problems of developing effective and safe treatments for rare variants appear formidable, as do the expense of developing treatments for large numbers of rare and unique genetic predispositions.</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Random select of an test example and visualisation of entities\n",
    "test_unit=random.choice(TEST_DATA)\n",
    "\n",
    "#print the existing entities (real Y)\n",
    "Y_entities =[]\n",
    "for i in test_unit[1]['entities']:\n",
    "    print(i[0],'->', i[1], i[2],':',test_unit[0][i[0]:i[1]])\n",
    "    Y_entities.append(({'start':i[0],\n",
    "     'end':i[1],\n",
    "     'label':i[2],\n",
    "     'name':test_unit[0][i[0]:i[1]]}))\n",
    "    \n",
    "#visualize the detected entities in this example \n",
    "from spacy import displacy\n",
    "doc2 = nlp_test(test_unit[0])\n",
    "displacy.render(doc2, style='ent', jupyter=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get the pre calculated entities, and see if they were detected by the Spacy model.\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "RES_TEST=[]\n",
    "id_doc=0\n",
    "for test_unit in TEST_DATA:    \n",
    "    id_doc+=1\n",
    "    doc2 = nlp_test(test_unit[0])\n",
    "    #create a dataframe of the existing entities \n",
    "    for i in test_unit[1]['entities']:\n",
    "        tmp = [id_doc,\n",
    "               i[0],\n",
    "               i[1],\n",
    "               i[2],\n",
    "               test_unit[0][i[0]:i[1]],\n",
    "               \"NOT_DETECTED\"]#put by default the label \"NOT_DETECTED\", if they was detected wil update the field\n",
    "        #test for each existing entities if it was detected by the model\n",
    "        for ent in doc2.ents:\n",
    "            if (tmp[1]==ent.start_char and tmp[2]==ent.end_char and tmp[3]==ent.label_ ):\n",
    "                tmp[5]=ent.label_ #update the \"NOT_DETECTED\" label by the one detected\n",
    "                        \n",
    "        RES_TEST.append(tmp)\n",
    "                    \n",
    "RES_TEST = pd.DataFrame(RES_TEST,columns=['id_doc', 'start', 'end', 'label', 'name', 'Y_hat_label'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Y_hat</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENE</th>\n",
       "      <th>NOT_DETECTED</th>\n",
       "      <th>rate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DISEASE</th>\n",
       "      <td>88</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0.838095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <td>0</td>\n",
       "      <td>328</td>\n",
       "      <td>953</td>\n",
       "      <td>0.256050</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Y_hat    DISEASE  GENE  NOT_DETECTED      rate\n",
       "Y                                             \n",
       "DISEASE       88     0            17  0.838095\n",
       "GENE           0   328           953  0.256050"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "table1 = pd.crosstab(RES_TEST['label'],RES_TEST['Y_hat_label'], rownames=['Y'], colnames=['Y_hat'])\n",
    "tmp =  table1['DISEASE'] + table1['GENE'] \n",
    "table1['rate'] = tmp / (tmp+table1['NOT_DETECTED'])\n",
    "table1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Y_hat</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENE</th>\n",
       "      <th>NOT_DETECTED</th>\n",
       "      <th>rate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCC8</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACACA</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGT</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>0.130435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTR1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AIF1</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT3</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANK2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOB</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARNTL</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>0.684211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASTN1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATR</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alzheimer's disease</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Angelman syndrome</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aniridia</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Asperger syndrome</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autism spectrum disorder</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0.454545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autistic disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BAD</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA1</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BTG3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bardet-Biedl syndrome</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bipolar disorder</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bloom syndrome</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>45</td>\n",
       "      <td>0.021739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>0.137931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C4A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACNA1C</th>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.933333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CARS</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAT</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>0.033333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAF1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REN</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rett syndrome</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC25A14</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC6A3</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC6A4</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SOD1</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>0.142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STX8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Schizophrenia</th>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spinal muscular atrophy</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pneumonia</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCF7L2</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFA</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFB1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFB2</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "      <td>0.014085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TLR4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNF</th>\n",
       "      <td>0</td>\n",
       "      <td>44</td>\n",
       "      <td>95</td>\n",
       "      <td>0.316547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TOX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRANK1</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSC1</th>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>0.916667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSC2</th>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>0.944444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTR</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Teratoma</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZBTB20</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-Hodgkin lymphoma</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>obsessive-compulsive disorder</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>post-traumatic stress disorder</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pre-eclampsia</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substance-induced psychosis</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>145 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Y_hat                           DISEASE  GENE  NOT_DETECTED      rate\n",
       "NAME                                                                 \n",
       "ABCC8                                 0     1             0  1.000000\n",
       "ACACA                                 0     5             0  1.000000\n",
       "ACE                                   0     0             5  0.000000\n",
       "AGT                                   0     3            20  0.130435\n",
       "AGTR1                                 0     1             0  1.000000\n",
       "AIF1                                  0     3             0  1.000000\n",
       "AKT3                                  0     3             0  1.000000\n",
       "ANK2                                  0     0             6  0.000000\n",
       "APOB                                  0     0             4  0.000000\n",
       "ARNTL                                 0    13             6  0.684211\n",
       "ASTN1                                 0     1             0  1.000000\n",
       "ATR                                   0     0             3  0.000000\n",
       "Alzheimer's disease                   7     0             0  1.000000\n",
       "Angelman syndrome                     0     0             1  0.000000\n",
       "Aniridia                              1     0             0  1.000000\n",
       "Asperger syndrome                     2     0             0  1.000000\n",
       "Autism spectrum disorder              5     0             6  0.454545\n",
       "Autistic disorder                     1     0             0  1.000000\n",
       "BAD                                   0     0             3  0.000000\n",
       "BRCA1                                 0     2             0  1.000000\n",
       "BTG3                                  0     0             1  0.000000\n",
       "Bardet-Biedl syndrome                 1     0             0  1.000000\n",
       "Bipolar disorder                      2     0             0  1.000000\n",
       "Bloom syndrome                        1     0             0  1.000000\n",
       "C2                                    0     1            45  0.021739\n",
       "C3                                    0     4            25  0.137931\n",
       "C4A                                   0     0             9  0.000000\n",
       "CACNA1C                               0    14             1  0.933333\n",
       "CARS                                  0    13             0  1.000000\n",
       "CAT                                   0     1            29  0.033333\n",
       "...                                 ...   ...           ...       ...\n",
       "RAF1                                  0     0             1  0.000000\n",
       "REN                                   0     0            15  0.000000\n",
       "Rett syndrome                         3     0             0  1.000000\n",
       "SLC25A14                              0     0             7  0.000000\n",
       "SLC6A3                                0     2             0  1.000000\n",
       "SLC6A4                                0     1             0  1.000000\n",
       "SOD1                                  0     3            18  0.142857\n",
       "STX8                                  0     0             3  0.000000\n",
       "Schizophrenia                        12     0             0  1.000000\n",
       "Spinal muscular atrophy               1     0             0  1.000000\n",
       "Streptococcus pneumonia               0     0             1  0.000000\n",
       "TCF7L2                                0     3             0  1.000000\n",
       "TGFA                                  0     1             0  1.000000\n",
       "TGFB1                                 0     1             0  1.000000\n",
       "TGFB2                                 0     4             0  1.000000\n",
       "TH                                    0     1            70  0.014085\n",
       "TLR4                                  0     0             1  0.000000\n",
       "TNF                                   0    44            95  0.316547\n",
       "TOX                                   0     0             1  0.000000\n",
       "TRANK1                                0     2             0  1.000000\n",
       "TSC1                                  0    11             1  0.916667\n",
       "TSC2                                  0    17             1  0.944444\n",
       "TTR                                   0     1             0  1.000000\n",
       "Teratoma                              0     0             1  0.000000\n",
       "ZBTB20                                0     0             1  0.000000\n",
       "non-Hodgkin lymphoma                  2     0             0  1.000000\n",
       "obsessive-compulsive disorder         0     0             1  0.000000\n",
       "post-traumatic stress disorder        2     0             0  1.000000\n",
       "pre-eclampsia                         1     0             0  1.000000\n",
       "substance-induced psychosis           1     0             0  1.000000\n",
       "\n",
       "[145 rows x 4 columns]"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table2 = pd.crosstab(RES_TEST['name'],RES_TEST['Y_hat_label'], rownames=['NAME'], colnames=['Y_hat'])\n",
    "\n",
    "tmp =  table2['DISEASE'] + table2['GENE'] \n",
    "table2['rate'] = tmp / (tmp+table2['NOT_DETECTED'])\n",
    "table2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
